packages/pds/src/api/com/atproto/admin/takeModerationAction.ts

```diff
@@ -1,16 +1,74 @@
 import { CID } from 'multiformats/cid'
 import { AtUri } from '@atproto/uri'
+import { isRepoRef } from '@atproto/api/src/client/types/com/atproto/admin/defs'
+import { isMain as isStrongRef } from '@atproto/api/src/client/types/com/atproto/repo/strongRef'
 import { Server } from '../../../../lexicon'
 import AppContext from '../../../../context'
 import { TAKEDOWN } from '../../../../lexicon/types/com/atproto/admin/defs'
 import { getSubject, getAction } from '../moderation/util'
 import { AuthRequiredError, InvalidRequestError } from '@atproto/xrpc-server'
+import { authPassthru } from './util'
 
 export default function (server: Server, ctx: AppContext) {
   server.com.atproto.admin.takeModerationAction({
     auth: ctx.moderatorVerifier,
-    handler: async ({ input, auth }) => {
+    handler: async ({ req, input, auth }) => {
       const { db, services } = ctx
+      if (ctx.shouldProxyModeration()) {
+        const { data: result } =
+          await ctx.appviewAgent.com.atproto.admin.takeModerationAction(
+            input.body,
+            authPassthru(req, true),
+          )
+
+        const transact = db.transaction(async (dbTxn) => {
+          const authTxn = services.auth(dbTxn)
+          const moderationTxn = services.moderation(dbTxn)
+          const labelTxn = services.appView.label(dbTxn)
+          // perform takedowns
+          if (result.action === TAKEDOWN && isRepoRef(result.subject)) {
+            await authTxn.revokeRefreshTokensByDid(result.subject.did)
+            await moderationTxn.takedownRepo({
+              takedownId: result.id,
+              did: result.subject.did,
+            })
+          }
+          if (result.action === TAKEDOWN && isStrongRef(result.subject)) {
+            await moderationTxn.takedownRecord({
+              takedownId: result.id,
+              uri: new AtUri(result.subject.uri),
+              blobCids: result.subjectBlobCids.map((cid) => CID.parse(cid)),
+            })
+          }
+          // apply label creation & negations
+          const applyLabels = (uri: string, cid: string | null) =>
+            labelTxn.formatAndCreate(ctx.cfg.labelerDid, uri, cid, {
+              create: result.createLabelVals,
+              negate: result.negateLabelVals,
+            })
+          if (isRepoRef(result.subject)) {
+            await applyLabels(result.subject.did, null)
+          }
+          if (isStrongRef(result.subject)) {
+            await applyLabels(result.subject.uri, result.subject.cid)
+          }
+        })
+
+        try {
+          await transact
+        } catch (err) {
+          req.log.error(
```

Comment:
what's the recovery from this?

now redsky will show that something is taken down but that isn't reflected on the PDS, right?

this is probably rare, but we should have a plan in mind for how to deal with it 🤔 
would it make sense to fire of a reversal if the txn fails on the PDS? to attempt to keep state in sync? (of course that request can fail as well 😅

Reply:
It's a good question. I was imagining we'd monitor these errors closely and handle them in a high-touch way.  My hunch is that if this happens it's probably not a temporary issue, and something needs fixing.